packages/r2/src/multipart.ts

```diff
@@ -0,0 +1,568 @@
+// We should aim to improve this in Miniflare 3.
+
+import assert from "assert";
+import { Blob } from "buffer";
+import crypto from "crypto";
+import { arrayBuffer } from "stream/consumers";
+import { ReadableStream } from "stream/web";
+import { TextEncoder } from "util";
+import { _isBodyStream, _isFixedLengthStream } from "@miniflare/core";
+import {
+  ParsedRange,
+  RequestContext,
+  Storage,
+  assertInRequest,
+  getRequestContext,
+  viewToArray,
+  waitForOpenInputGate,
+  waitForOpenOutputGate,
+} from "@miniflare/shared";
+import {
+  MAX_KEY_SIZE,
+  R2MultipartReference,
+  R2Object,
+  R2ObjectMetadata,
+  createMD5Hash,
+  createVersion,
+} from "./r2Object";
+
+/** @internal */
+export const _INTERNAL_PREFIX = "__MINIFLARE_INTERNAL__";
+
+const MIN_MULTIPART_UPLOAD_SIZE = 5 * 1024 * 1024;
+
+export interface R2UploadedPart {
+  partNumber: number;
+  etag: string;
+}
+export interface R2MultipartPendingIndexMetadata {
+  httpMetadata: R2ObjectMetadata["httpMetadata"];
+  customMetadata: R2ObjectMetadata["customMetadata"];
+}
+export type R2MultipartIndexMetadata =
+  | R2MultipartPendingIndexMetadata
+  | { aborted: true }
+  | { completed: true };
+interface R2MultipartPartMetadata {
+  size: number;
+  md5: string;
+  etag: string;
+}
+
+type R2UploadState =
+  | { exists: true; meta: R2MultipartPendingIndexMetadata }
+  | { exists: false; aborted: boolean; completed: boolean };
+
+const encoder = new TextEncoder();
+export function validateMultipartKey(method: string, key: string) {
+  if (
+    Buffer.byteLength(key) > MAX_KEY_SIZE ||
+    key.startsWith(_INTERNAL_PREFIX)
+  ) {
+    throw new TypeError(
+      `${method}: The specified object name is not valid. (10020)`
+    );
+  }
+}
+function validatePartNumber(partNumber: number) {
+  if (partNumber >= 1 && partNumber <= 10_000) return;
+  throw new TypeError(
+    `Part number must be between 1 and 10000 (inclusive). Actual value was: ${partNumber}`
+  );
+}
+
+function generateId(likelyOnFilesystem = false) {
+  // Windows has a maximum path length of ~260 characters:
+  // https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation
+  // Miniflare R2 buckets will usually be backed by file-system storage,
+  // especially when using multipart uploads for large files. Therefore, reduce
+  // the size of the upload ID on Windows, preferring a longer ID otherwise
+  // to more closely match R2 behaviour.
+  const size = likelyOnFilesystem && process.platform === "win32" ? 32 : 128;
+  return crypto.randomBytes(size).toString("base64url");
+}
+function generateMultipartEtag(md5Hexes: string[]) {
+  // TODO: R2's multipart ETags don't seem to be deterministic, should ours be?
+  // https://stackoverflow.com/a/19896823
+  const hash = crypto.createHash("md5");
+  for (const md5Hex of md5Hexes) hash.update(Buffer.from(md5Hex, "hex"));
+  return `${hash.digest("hex")}-${md5Hexes.length}`;
```

Comment:
Assuming that this is how S3 calculates their etags, we might move to this way of calculating them ourselves in the future.